### PR TITLE
Add test for lock implementation in message_switch

### PR DESCRIPTION
**Pull request by Vincent-lau**
The lock implementation for `protocol_async` was not correct, hence adding the test suggested by Edwin to trigger race.

Test works by trying to acquire a standard Ocaml mutex in a critical section provided by the `with_lock` implementation. This means that any failed attempt to acquire the lock implies the `with_lock` implementation does not provide mutual exclusion.

### CODE REVIEW

**Comment by Vincent-lau** [comment]
Related fix for the race condition in async #5359

**Comment by Vincent-lau** [comment]
Some quality gate issues regarding ml files. Do we care about ml files in `core_test`? We can either add path to exclude or add mli files for these tests.

File: ocaml/message-switch/core/mtest.ml
Comment (edwintorok) [inline ocaml/message-switch/core/mtest.ml:27]: For Lwt I think we'll need to wait until all of these are complete. Instead of ignoring the array try storing it in a value (or as a list).

File: ocaml/message-switch/core/mtest.ml
Comment (edwintorok) [inline ocaml/message-switch/core/mtest.ml:25]: Move this after the array creation: then you don't need to sleep for 1s, you can immediately broadcast after the array has been created.

File: ocaml/message-switch/core/mtest.ml
Comment (edwintorok) [inline ocaml/message-switch/core/mtest.ml:39]: would be good to wait until all promises in the array are resolved to ensure we've let the code run fully. Then we can remove the arbitrary 2s sleep, and make this test complete more quickly.

**Review by edwintorok — COMMENTED** [review]
Thanks, this test is a good start. See my comments for some suggestions on how to eliminate the sleeps and make the test more robust for Lwt.

**Comment by lindig** [comment]
Are we not trying to remove all Async code and is this adding more?

**Comment by psafont** [comment]
It's adding a test to ensure the async code works as expected. This will be removed in the coming months, but I believe it's good we have proof of the fix in the repo

**Comment by psafont** [comment]
Please rebase, tests should be more stable now

**Review by psafont — APPROVED** [review]

**Comment by lindig** [comment]
@edwintorok you had some initial comments; so would suggest to revisit :arrow_double_up:

**Review by lindig — APPROVED** [review]
Looks good to me but I think Edvin should take a look because he had comments.

**Comment by Vincent-lau** [comment]
We can probably merge this one if @edwintorok is busy, it's a small change in unit tests anyway.

**Comment by psafont** [comment]
I agree, this is low risk

**Pull request merged**
